pretrain/redrock_trainer.py

```diff
@@ -81,18 +82,67 @@ def __init__(
     self.backward_nvtx_range = None
     self.max_iters = max_iters
     self.warmup_iters = warmup_iters
+    self.fast_init = fast_init
     self.trainer = trainer
 
   def configure_model(self) -> None:
     import psutil
     import time
     print(self.trainer.global_rank, ' in configure_model', flush=True)
     print('RAM Used (GB):', psutil.virtual_memory()[3]/1000000000, flush=True)
-    t = time.time()
-    self.module = GPT(self.config)
-    print(f'{self.trainer.global_rank} time to init model: {(time.time()-t):.02f}s', flush=True)
-    t = time.time()
-    self.module.apply(self.module._init_weights)
+    if self.fast_init:
+      t = time.time()
+      # init meta model
+      with self.trainer.init_module(empty_init=True):
+        self.module = GPT(self.config)
+      
+      # Iterate through each module in the model, replacing meta layers
+      # with real layers after initializing them and their weights.
+      # All layers not initialized below will be initialized later by FSDP
+      for name, module in self.module.named_modules():
+        state_dict = {}
+        if isinstance(module, torch.nn.Linear):
+          # define new layer on cuda so weight initialization is much faster
+          with torch.device('cuda'):
+            new_linear = torch.nn.Linear(
+                module.in_features,
+                module.out_features,
+                bias=True if module.bias is not None else False
+            )
+
+          # initialize weights
+          new_linear.apply(self.module._init_weights)
+          
+          # move new layer to cpu & prepare to load into model
+          new_linear.to('cpu')
+          state_dict[f"{name}.weight"] = new_linear.weight
+
+          if module.bias is not None:
+            state_dict[f"{name}.bias"] = new_linear.bias
+
+          elif isinstance(module, torch.nn.Embedding):
```

Comment:
Is the indentation to be aligned with `if isinstance(module, torch.nn.Linear):` above?

Reply:
yes, thank you for catching this. I must have made this mistake when copying over code.